zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java

```diff
@@ -1011,15 +1012,29 @@ public static List<ACL> fixupACL(String path, List<Id> authInfo, List<ACL> acls)
             if (id == null || id.getScheme() == null) {
                 throw new KeeperException.InvalidACLException(path);
             }
-            if (id.getScheme().equals("world") && id.getId().equals("anyone") && !X509AuthenticationConfig
-                .getInstance().isX509ClientIdAsAclEnabled()) {
+            boolean isX509ZnodeGroupAclSuperUser =
+                X509AuthenticationConfig.getInstance().isX509ClientIdAsAclEnabled() && authInfo
+                    .stream().anyMatch(authInfoId ->
+                        authInfoId.getScheme().equals(X509AuthenticationUtil.SUPERUSER_AUTH_SCHEME)
+                            && authInfoId.getId().equals(
+                            X509AuthenticationConfig.getInstance().getZnodeGroupAclSuperUserId()));
+            if ((id.getScheme().equals("world") && id.getId().equals("anyone") && !X509AuthenticationConfig
+                .getInstance().isX509ClientIdAsAclEnabled()) || isX509ZnodeGroupAclSuperUser) {
                 rv.add(a);
             } else if (id.getScheme().equals("auth") || X509AuthenticationConfig
                 .getInstance().isX509ClientIdAsAclEnabled()) {
                 // This is the "auth" id, so we have to expand it to the
                 // authenticated ids of the requestor
                 boolean authIdValid = false;
                 for (Id cid : authInfo) {
+
+                    // Special handling for cross domain component use cases when X509ClientIdAsAcl is enabled
+                    if (cid.getScheme().equals(X509AuthenticationUtil.SUPERUSER_AUTH_SCHEME)) {
+                        authIdValid = true;
+                        rv.add(new ACL(a.getPerms(), new Id("x509", cid.getId())));
```

Comment:
do we have a string constant for "x509"? If so, let's use that?

Reply:
There is currently no string constant for "x509", there are a few places using this string right now. I would break this into a separate PR for a refactor later.